### PR TITLE
Added version_tree module to support versioning of arbitrary objects …

### PR DESCRIPTION
**Pull request by vbawa**
…in IPFS

### CODE REVIEW

**Comment by vbawa** [comment]
@iamtrask, here's the initial implementation of a versioning API for IPFS that we were discussing: it allows committing objects along with an IPFS parent hash (if not root), and retrieving the chain of ancestors from a node to its root.

I had to make some design and performance choices to get this implemented, so let me know if any of them don't fit the use-cases you have in mind.

A couple of things to take note of:
1) For the chain of ancestor methods, I'm using an iterator so that we can avoid cluttering the API with a limit parameter, but can still get arbitrary-length ancestor chains.
2) These methods also return the full contents of the chain of ancestors (in the form of VersionTreeNodes), instead of just their identifying multihashes.  Changing this should be trivial.
3) This isn't directly related, but I'm requiring an ipfsapi.Client to be passed in to either __init__ or each method. What do you think about having a singleton in the ipfsapi.Client class? It wouldn't require refactoring existing code and it means we can avoid passing around the object all over the place or creating it in multiple places.  The only reason I can think of that this wouldn't work would be if a given worker node wants to have multiple clients with different configurations for connecting to IPFS daemons.
4) I'm planning to write some unit tests but wanted to ask you if there's any guidelines I should be following for doing so.  I tested it manually end-to-end on IPFS and it seems to work.

**Comment by iamtrask** [comment]
This looks fantastic!! If I could make one request - would you be so kind as to create / include a jupyter notebook showing how you use it? This will greatly help others use your work. Please include it in (https://github.com/OpenMined/Grid/tree/master/notebooks/experimental)

**Comment by iamtrask** [comment]
I do also greatly enjoy your very well commented and highly readable code!

**Comment by vbawa** [comment]
Will do.  Any thoughts on the unit tests, or is the Jupyter notebook sufficient for now?

File: grid/ipfsapi/version_tree.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        # Convert empty string to None to minimize typing bugs.
+        self.parent_hash = parent_hash.decode("utf-8") or None
+        self.ipfs_client = ipfs_client
+
+    # TODO: Should there just be a singleton IPFS client in the ipfsapi.client
+    # module?  It seems like the current architecture would require
+    # it to either be passed around all over the place or potentially go
+    # through the connection process multiple times, and I don't think
+    # multiple configurations need to be maintained for a given worker node.
+    def commit(self, ipfs_client: ipfsapi.Client = None) -> str:
```
Comment (jvmancuso) [inline grid/ipfsapi/version_tree.py]: We'll want to expose this in other methods at some point. For example with torch, we'll usually want to commit gradients of model Parameters, either directly on the Parameter's grad or with some helper function.

**Comment by jvmancuso** [comment]
General notes -- great stuff!  Really like your coding style here, it's very well written.  I also like that this isn't specific to pubsub, torch, keras, etc.

As far as unit tests go, if you can provide some simple test cases in the notebook or a separate file, that'd be rad.  That way we can integrate those into a formal testing module when it's ready.  Whatever testing you did to verify on your own over IPFS should be good.

File: grid/ipfsapi/version_tree.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                 ipfs_client: ipfsapi.Client = None):
+        """ parent_hash is a UTF-8 IPFS multihash identifying
+        this node's parent in the version tree. If parent_hash is None,
+        this node is the root of a version tree. """
+        self.contents = contents
+        # Convert empty string to None to minimize typing bugs.
+        self.parent_hash = parent_hash.decode("utf-8") or None
+        self.ipfs_client = ipfs_client
+
+    # TODO: Should there just be a singleton IPFS client in the ipfsapi.client
```
[the remark this answers is not in the document]
Comment (vbawa) [reply grid/ipfsapi/version_tree.py]: > We can pass either the worker or the ipfs client from the worker into this class via ```__init__``` and refer to it as an attribute thereafter

This particular class is currently written to handle the IPFS client this way (along with optionally accepting it in each method that needs it).

> The ipfs client is already initialized via utils.get_ipfs_api in base_worker.py, where it becomes the ```api``` attribute.

RIght;  but my thoughts were that a reference to the Client will need to be passed through the stack wherever it's required (i.e. whenever the LoC we're on is not in the BaseWorker method so the self.api reference to the Client is not available).  As the code gets more complex, I figured a singleton reference might end up simplifying a bunch of APIs, and I think the functionality is well-suited to a singleont.  As long as a given worker will generally only have a single set of parameters that it wants to parametrize its IPFS connection with, I don't think this should present any issues.

> unless you had a strong preference for this?

No strong feelings about it;  it was more of a general thought I wanted to bring up before I forgot about it as opposed to anything directly relevant to this PR.

File: grid/ipfsapi/version_tree.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    # module?  It seems like the current architecture would require
+    # it to either be passed around all over the place or potentially go
+    # through the connection process multiple times, and I don't think
+    # multiple configurations need to be maintained for a given worker node.
+    def commit(self, ipfs_client: ipfsapi.Client = None) -> str:
+        """ Commits the node to the version tree, and returns the
+        UTF-8 multihash representing its IPFS ID"""
+        return (ipfs_client or self.ipfs_client).add_bytes(self.to_bytes())
+
+    @staticmethod
```
[the remark this answers is not in the document]
Comment (vbawa) [reply grid/ipfsapi/version_tree.py]: Done, converted all staticmethods to classmethods and removed all other references to VersionTreeNode in the method bodies.

File: grid/ipfsapi/version_tree.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        return VersionTreeNode.get_node_by_hash(
+            multihash, ipfs_client).get_with_ancestors(ipfs_client)
+
+    def to_bytes(self) -> bytes:
+        """ For contents b"foo", parent_hash "bar", and DELIMITER b"|",
+        returns b"foo|bar" """
+        parent_hash_bytes = self.parent_hash.encode("utf-8") if \
+            self.parent_hash else \
+            b""
+        return VersionTreeNode.DELIMITER.join(
```
[the remark this answers is not in the document]
Comment (vbawa) [reply grid/ipfsapi/version_tree.py]: Done

File: grid/ipfsapi/version_tree.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            (self.contents, parent_hash_bytes)
+        )
+
+    @staticmethod
+    def from_bytes(b: bytes) -> "VersionTreeNode":
+        return VersionTreeNode(*b.split(VersionTreeNode.DELIMITER))
```
[the remark this answers is not in the document]
Comment (vbawa) [reply grid/ipfsapi/version_tree.py]: Done

**Comment by vbawa** [comment]
The Jupyter notebook is about half-done;  Ideally, I'll push it sometime this weekend.

**Comment by yanndupis** [comment]
Hey @vbawa, I should have ping you yesterday that I was also working on a notebook. I am sure they will be complementary. Here is an example how we can version control weights of a model. 
https://github.com/yanndupis/Grid/blob/weights_version_control/notebooks/experimental/Weights_version_control.ipynb
This class is awesome, we could version control anything we want. I will push this notebook once your PR is merged.

**Comment by vbawa** [comment]
Alright, sorry for the delay.  I've pushed a Jupyter notebook with usage examples and some basic explanation.  @jvmancuso, let me know if there's anything else you'd like to see changed here.

**Comment by iamtrask** [comment]
This is a gorgeous Pull Request - absolutely love the generic support for tree storage in IPFS.

**Pull request merged**